lib/srv/discovery/discovery.go

```diff
@@ -1686,14 +1686,37 @@ func splitMatchers[T types.Matcher](matchers []T, matcherTypeCheck func(string)
 	return
 }
 
-func (s *Server) updatesEmptyDiscoveryGroup(getter func() (types.ResourceWithLabels, error)) bool {
-	if s.DiscoveryGroup == "" {
-		return false
+func (s *Server) resolveCreateErr(createErr error, discoveryOrigin string, getter func() (types.ResourceWithLabels, error)) error {
+	// We can only resolve the error if we have a discovery group configured
+	// and the error is that the resource already exists.
+	if s.DiscoveryGroup == "" || !trace.IsAlreadyExists(createErr) {
+		return trace.Wrap(createErr)
 	}
+
 	old, err := getter()
 	if err != nil {
-		return false
+		return trace.NewAggregate(createErr, err)
+	}
+
+	// Check that the registered resource origin matches the origin we want.
+	oldOrigin, err := types.GetOrigin(old)
+	if err != nil {
+		return trace.NewAggregate(createErr, err)
+	}
+	if oldOrigin != discoveryOrigin {
```

Comment:
We check if the old group is not empty below because that's a special case where the old/new groups don't match but we still want to update the resource.
In this way, discovery agents with a `discovery_group` essentially claim the resources they discover that used to be (or currently are) discovered by an agent that did not have the group configured.

For origin, we don't want that behavior - empty origin should not be updated because it's definitely not managed by discovery service.

Reply:
since anyone reading this code might make the same assumption (admittedly the intent here is subtle), I went ahead and added test cases for empty origin and a comment explaining the above as well.

Reply:
https://github.com/gravitational/teleport/compare/f1dd572a2b55425ac1ecb7e3e601cb1289144bde..1190fed6c6047125c3e0518736ef6c589086c9be

Reply:
Databases should always have an origin:
https://github.com/gravitational/teleport/blob/ca0c0f4fd9e844c1bf11105afe1c284bf78cf911/lib/auth/grpcserver.go#L3715-L3722